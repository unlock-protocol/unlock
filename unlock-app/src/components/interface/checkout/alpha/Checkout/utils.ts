interface FetchDataOptions {
  recipients: string[]
  network: number
  lockAddress: string
}

export async function fetchData(url: string, options: FetchDataOptions) {
  try {
    const endpoint = new URL(url)
    endpoint.searchParams.append('network', options.network.toString())
    endpoint.searchParams.append('lockAddress', options.lockAddress)

    for (const recipient of options.recipients) {
      endpoint.searchParams.append('recipients[]', recipient)
    }

    const abortController = new AbortController()

    const timer = setTimeout(() => {
      abortController.abort()
    }, 5000)

    const response = await fetch(endpoint, {
      headers: {
        'Content-Type': 'application/json',
      },
      signal: abortController.signal,
    })

    clearTimeout(timer)

    if (!response.ok) {
      throw new Error('Failed to fetch data.')
    }

    const json = await response.json()
    return json.data as string[]
  } catch {
    return null
  }
}
