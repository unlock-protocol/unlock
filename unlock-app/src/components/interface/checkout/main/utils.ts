import { PaywallLockConfigType } from '@unlock-protocol/core'
import { PaywallConfig } from '~/unlockTypes'
interface FetchRecipientsDataOptions {
  recipients: string[]
  network: number
  lockAddress: string
}

export async function fetchRecipientsData(
  url: string,
  { lockAddress, network, recipients }: FetchRecipientsDataOptions
) {
  try {
    const result: string[] = []
    for (const recipient of recipients) {
      const dataEndpoint = new URL(url)

      dataEndpoint.searchParams.append('network', network.toString())
      dataEndpoint.searchParams.append('lockAddress', lockAddress)
      dataEndpoint.searchParams.append('recipient', recipient)

      // We need to proxy to avoid cors.
      const endpoint = new URL('/data', 'https://rpc.unlock-protocol.com')
      endpoint.searchParams.append('url', dataEndpoint.toString())

      const abortController = new AbortController()

      const timer = setTimeout(() => {
        abortController.abort()
      }, 5000)

      const response = await fetch(endpoint, {
        headers: {
          'Content-Type': 'application/json; charset=UTF-8',
        },
        signal: abortController.signal,
      })

      clearTimeout(timer)

      if (!response.ok) {
        throw new Error('Failed to fetch data.')
      }

      const json = await response.json()
      result.push(json.data)
      return result
    }
  } catch (error) {
    console.error(error)
    return
  }
}

export function formatFiatPriceFromCents(priceInCents: number, currency = '$') {
  const formatted = (priceInCents / 100).toLocaleString(undefined, {
    maximumFractionDigits: 2,
    minimumFractionDigits: 2,
  })
  return `${currency} ${formatted}`
}
interface Options {
  paywallConfig?: PaywallConfig
  lock?: PaywallLockConfigType
}

export const shouldSkip = ({ lock, paywallConfig }: Options) => {
  const maxRecipients = lock?.maxRecipients || paywallConfig?.maxRecipients
  const minRecipients = lock?.minRecipients || paywallConfig?.minRecipients
  const hasMaxRecipients = maxRecipients && maxRecipients > 1
  const hasMinRecipients = minRecipients && minRecipients > 1
  const skipQuantity = !(hasMaxRecipients || hasMinRecipients)

  const skip = lock?.skipRecipient || paywallConfig?.skipRecipient
  const collectsMetadadata =
    lock?.metadataInputs ||
    paywallConfig?.metadataInputs ||
    paywallConfig?.emailRequired ||
    lock?.emailRequired
  const skipRecipient = Boolean(skip && !collectsMetadadata)

  return {
    skipQuantity,
    skipRecipient,
  }
}
