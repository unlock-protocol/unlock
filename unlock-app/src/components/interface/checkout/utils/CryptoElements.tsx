// https://stripe.com/docs/crypto/quickstart

import { createContext, useEffect, useState, useRef, useContext } from 'react'

interface CryptoElementsContext {
  onramp: any
}

// ReactContext to simplify access of StripeOnramp object
const CryptoElementsContext = createContext<CryptoElementsContext | undefined>(
  undefined
)
CryptoElementsContext.displayName = 'CryptoElementsContext'

interface CryptoElementsProps {
  stripeOnramp: any
  children: React.ReactNode
}

export const CryptoElements = ({
  stripeOnramp,
  children,
}: CryptoElementsProps) => {
  const [ctx, setContext] = useState(() => ({
    onramp: null,
  }))

  useEffect(() => {
    let isMounted = true

    Promise.resolve(stripeOnramp).then((onramp) => {
      if (onramp && isMounted) {
        setContext((ctx) => (ctx.onramp ? ctx : { onramp }))
      }
    })

    return () => {
      isMounted = false
    }
  }, [stripeOnramp])

  return (
    <CryptoElementsContext.Provider value={ctx}>
      {children}
    </CryptoElementsContext.Provider>
  )
}

// React hook to get StripeOnramp from context
export const useStripeOnramp = () => {
  const context = useContext(CryptoElementsContext)
  return context?.onramp
}

// React element to render Onramp UI
const useOnrampSessionListener = (
  type: string,
  session: any,
  callback: any
) => {
  useEffect(() => {
    if (session && callback) {
      const listener = (e: any) => callback(e.payload)
      session.addEventListener(type, listener)
      return () => {
        session.removeEventListener(type, listener)
      }
    }
    return () => {}
  }, [session, callback, type])
}

interface OnrampElementProps {
  clientSecret: string
  appearance: any
  onChange: (payload: any) => void
}

export const OnrampElement = ({
  clientSecret,
  appearance,
  onChange,
  ...props
}: OnrampElementProps) => {
  const stripeOnramp = useStripeOnramp()
  const onrampElementRef = useRef<HTMLDivElement | null>(null)
  const [session, setSession] = useState()

  const appearanceJSON = JSON.stringify(appearance)
  useEffect(() => {
    const containerRef = onrampElementRef.current
    if (containerRef) {
      // NB: ideally we want to be able to hot swap/update onramp iframe
      // This currently results a flash if one needs to mint a new session when they need to udpate fixed transaction details
      containerRef.innerHTML = ''

      if (clientSecret && stripeOnramp) {
        setSession(
          stripeOnramp
            .createSession({
              clientSecret,
              appearance: appearanceJSON ? JSON.parse(appearanceJSON) : {},
            })
            .mount(containerRef)
        )
      }
    }
  }, [appearanceJSON, clientSecret, stripeOnramp])

  useOnrampSessionListener('onramp_session_updated', session, onChange)

  return <div {...props} ref={onrampElementRef}></div>
}
