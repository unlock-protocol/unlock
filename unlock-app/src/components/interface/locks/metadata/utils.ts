export interface Ticket {
  event_start_date?: string
  event_start_time?: string
  event_address?: string
  event_meeting_url?: string
}
export interface MetadataFormData {
  name: string
  description?: string
  external_url?: string
  youtube_url?: string
  animation_url?: string
  background_color?: string
  ticket: Ticket
  properties: Attribute[]
  levels: Attribute[]
  stats: Attribute[]
  [key: string]: any
}

export interface Attribute {
  display_type?: string
  max_value?: number
  trait_type: string
  value: string | number
}

export interface Metadata {
  name: string
  image?: string
  description?: string
  external_url?: string
  youtube_url?: string
  animation_url?: string
  background_color?: string
  attributes?: Attribute[]
  [key: string]: any
}

export function toFormData({
  name,
  description,
  attributes,
  animation_url,
  external_url,
  youtube_url,
  background_color,
}: Metadata) {
  const ticket = attributes
    ?.filter((item) => item.trait_type.startsWith('event_'))
    .reduce((item, { trait_type, value }) => {
      item[trait_type as keyof Ticket] = value as string
      return item
    }, {} as Ticket)

  const stats = attributes?.filter(
    (item) => item.display_type === 'number' && typeof item.value === 'number'
  )
  const levels = attributes?.filter(
    (item) => typeof item.value === 'number' && !item.display_type
  )
  const properties = attributes?.filter(
    (item) =>
      typeof item.value === 'string' &&
      !item.max_value &&
      !item.trait_type.startsWith('event_')
  )
  return {
    name,
    description,
    background_color,
    animation_url,
    external_url,
    youtube_url,
    ticket,
    levels,
    properties,
    stats,
  } as MetadataFormData
}
