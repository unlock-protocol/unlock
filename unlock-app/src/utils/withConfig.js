import React from 'react'
import configure from '../config'
import { WrongNetwork, MissingProvider } from '../components/creator/FatalError'
import { ETHEREUM_NETWORKS_NAMES } from '../constants'
import { connect } from 'react-redux'

/**
 * Function which creates higher order component with the config
 * Taken from https://reactjs.org/docs/context.html#consuming-context-with-a-hoc
 */

const isServer = typeof window === 'undefined'
const config = configure(global)
const ConfigContext = React.createContext(config)

// This function takes a component...
export function withConfig(Component) {
  // ...and returns another component...
  function componentWithConfig(props) {
    // ... and renders the wrapped component with the context config!
    // Notice that we pass through any additional props as well

    const { store: reduxStore, router } = props

    if (!isServer) {
      // Ensuring that we have at least a provider
      if (Object.keys(config.providers).length === 0) {
        return (<MissingProvider />)
      }

      // Ensuring that the provider is using the right network!
      if (router.route !== '/provider' && config.isRequiredNetwork && !config.isRequiredNetwork(reduxStore.getState().network.name)) {
        return (<WrongNetwork currentNetwork={ETHEREUM_NETWORKS_NAMES[reduxStore.getState().network.name][0]} requiredNetwork={config.requiredNetwork} />)
      }
    }

    return (
      <ConfigContext.Consumer>
        {config => <Component {...props} config={config} />}
      </ConfigContext.Consumer>
    )
  }

  function mapStateToProps (state) {
    return {
      account: state.network.account, // TODO change account to base level
      network: state.network,
    }
  }

  componentWithConfig.getInitialProps = async context => {
    return {
      ...(Component.getInitialProps ? await Component.getInitialProps(context) : {}),
    }
  }

  return connect(mapStateToProps)(componentWithConfig)
}
