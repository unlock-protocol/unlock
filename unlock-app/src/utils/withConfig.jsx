import React, { useContext } from 'react'

/**
 * Function which creates higher order component with the config
 * Taken from https://reactjs.org/docs/context.html#consuming-context-with-a-hoc
 */

export const ConfigContext = React.createContext()

/**
 * This creates an HOC from a component and injects the configuration.
 * It also triggers errors if constraints are not respected.
 * @param {*} Component
 */
export default function withConfig(Component) {
  function componentWithConfig(props) {
    return (
      <ConfigContext.Consumer>
        {(config) => <Component {...props} config={config} />}
      </ConfigContext.Consumer>
    )
  }

  return componentWithConfig
}

export function useConfig() {
  return useContext(ConfigContext)
}
