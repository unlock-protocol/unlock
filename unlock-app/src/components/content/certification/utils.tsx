import { Metadata } from '~/components/interface/locks/metadata/utils'

interface CertificationUrlProps {
  metadata?: Partial<Metadata>
  lockAddress: string
  network: string | number
  tokenId?: string | number
}

export const getCertificationPath = ({
  metadata,
  lockAddress,
  network,
  tokenId,
}: CertificationUrlProps): string => {
  const slug = metadata?.slug

  if (slug) {
    return `/certification?s=${slug}`
  }

  if (tokenId) {
    return `/certification?lockAddress=${lockAddress}&network=${network}&tokenId=${tokenId}`
  }

  return `/certification?lockAddress=${lockAddress}&network=${network}`
}
