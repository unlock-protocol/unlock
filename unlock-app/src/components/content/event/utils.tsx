import dayjs from 'dayjs'
import utc from 'dayjs/plugin/utc'
import timezone from 'dayjs/plugin/timezone'
import { Metadata } from '~/components/interface/locks/metadata/utils'

dayjs.extend(utc)
dayjs.extend(timezone)

export const getEventDate = (ticket: any): Date | null => {
  if (ticket?.event_start_date) {
    const timestamp = [ticket.event_start_date, ticket.event_start_time].join(
      ' '
    )
    const dayjsLocal = dayjs.tz(timestamp, ticket.event_timezone)
    return dayjsLocal.toDate()
  }

  return null
}

export const getEventEndDate = (ticket: any): Date | null => {
  if (ticket?.event_end_date) {
    const timestamp = [ticket.event_end_date, ticket.event_end_time].join(' ')

    const dayjsLocal = dayjs.tz(timestamp, ticket.event_timezone)
    return dayjsLocal.toDate()
  }

  return null
}

interface EventUrlProps {
  metadata?: Partial<Metadata>
  lockAddress: string
  network: string | number
}

export const getEventPath = ({
  metadata,
  lockAddress,
  network,
}: EventUrlProps): string => {
  const slug = metadata?.slug

  if (slug) {
    return `/event/${slug}`
  }
  return `/event?lockAddress=${lockAddress}&network=${network}`
}

export const getEventUrl = ({
  metadata,
  lockAddress,
  network,
}: EventUrlProps): string => {
  if (typeof window !== 'undefined' && window?.location?.origin) {
    return `${window.location.origin}${getEventPath({
      metadata,
      lockAddress,
      network,
    })}`
  }
  return getEventPath({
    metadata,
    lockAddress,
    network,
  })
}
