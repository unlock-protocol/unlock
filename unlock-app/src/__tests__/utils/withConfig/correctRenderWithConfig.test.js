import React from 'react'
import renderer from 'react-test-renderer'
import { withConfig } from '../../../utils/withConfig'

const state = {
  account: {
    address: '0xdeadbeef',
    balance: '1000',
  },
  network: {
    name: 4,
  },
}

const store = {
  getState: jest.fn(() => state),
  dispatch: jest.fn(),
  subscribe: jest.fn(),
}

const Component = () =>
  <div>
    An unlock component
  </div>

const ComponentWithConfig = withConfig(Component)

describe('withConfig High Order Component', () => {
  it('should render correctly', () => {
    const tree = renderer
      .create(<ComponentWithConfig store={store} router={{ route: '/provider' }} />)
      .toJSON()
    expect(tree).toMatchSnapshot()
  })
})
