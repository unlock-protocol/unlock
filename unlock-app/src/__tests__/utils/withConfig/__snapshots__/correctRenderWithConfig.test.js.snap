// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`withConfig High Order Component should render correctly 1`] = `
<div>
  An unlock component
</div>
`;
